docs/markdown/Ad-Hoc Tools/adhoc-tool.md

```diff
@@ -0,0 +1,86 @@
+
+
+### Specifying dependencies
+
+`adhoc_tool` has more complexity surrounding dependencies compared with Pants' first-class targets. This is because you need to do manual work to set up the execution environment, which is usually taken care of by plugin code.
+
+`adhoc_tool` has three dependencies fields:
+
+* `output_dependencies`, which defines dependencies that are required to effectively consume the output of the tool, _e.g._ runtime libraries for generated code bindings. These are returned when resolving the transitive dependencies of any (transitive) dependents of the `adhoc_tool` target.
+* `execution_dependencies`, which define data dependencies required for the tool to produce its output. These are not considered when resolving transitive dependencies that include this `adhoc_tool` target.
+* `runnable_dependencies`, which define runnables that the `adhoc_tool` needs to execute as a subprocess. These are also not considered when resolving transitive dependencies. The discussion of `system_binary` later in this page shows one key use of `runnable_dependencies`.
+
+In the `antlr` example, `output_dependencies` is used because the tool produces Python-based bindings that depend on a runtime library. `execution_dependencies` specifies the sources that are consumed by the tool, but do not need to be consumed by subsequent build steps.
+
+
+### Specifying outputs
+
+Generally, `adhoc_tool` targets are run to produce outputs that can be supplied to other targets. These can be in the form of files or directories that are output directly by the tools: use the `output_files` field to capture individual files, or `output_directories` to capture entire directories as output. 
+
+Files are captured relative to the build root by default: this is useful when passing results to further `adhoc_tool` targets defined in the same `BUILD` file. If this behavior is not right for you, for example, if you are producing an artifact for packaging, you can change the root of the outputs using the `root_output_directory` field.
+
+Finally, if you want to capture `stdout` or `stderr` from your tool, you can use the `stdout` or `stderr` fields. These specify filenames where those streams will be dumped once the process completes.
+
+
+### Chaining processes together
+
+_(Our [JavaScript demo](https://github.org/pantsbuild/example-adhoc/tree/main/javascript)) demonstrates a string of `adhoc_tool` targets that's used to produce a resource file._
+
+To get the best cache efficiency, it can make sense to break your `adhoc_tool` into smaller incremental steps. For example, if your process needs to fetch dependencies and then build a library based on those dependencies and some first-party source files, having one `adhoc_tool` for each of those steps means that the dependency-fetching stage will only be re-run when your requirements change, and not when the first-party source files change.
```

Comment:
Is it worth talking about the caching behaviour in more detail: e.g. that similar to `shell_command`, these should be "pure" processes, where their only observable output is the files on disk? And that each `adhoc_tool` is cached based on its configuration and input files, meaning non-deterministic invocations (e.g. something that calls `date`/`datetime.datetime.now()`/`new Date()`/...) might give unexpected or inconsistent behaviour?

Maybe not here, but it feels like the (good) recommendations here would "click" better for users with a slightly deeper understanding of the cache moving parts.